src/providers/googleProvider.ts

```
@@ -1,59 +1,119 @@
+// this to make it as easy as possible for the end user?
+// We could perhaps instruct them to call loadGeocoder() and
+// pass the results as the second arg... or wrap it all in a promise
+// and return that instead?
 export default class GoogleProvider extends AbstractProvider<
   RequestResult,
-  RawResult
+  google.maps.GeocoderResult
 > {
-  searchUrl = 'https://maps.googleapis.com/maps/api/geocode/json';
+  geoCoder: google.maps.Geocoder;
 
-  endpoint({ query }: EndpointArgument) {
-    const params = typeof query === 'string' ? { address: query } : query;
-    return this.getUrl(this.searchUrl, params);
+  constructor(geoCoder: google.maps.Geocoder) {
+    super();
+    this._checkGeoCoder(geoCoder);
+    this.geoCoder = geoCoder;
   }
 
-  parse(result: ParseArgument<RequestResult>): SearchResult<RawResult>[] {
-    return result.data.results.map((r) => ({
-      x: r.geometry.location.lng,
-      y: r.geometry.location.lat,
-      label: r.formatted_address,
-      bounds: [
-        [r.geometry.viewport.southwest.lat, r.geometry.viewport.southwest.lng], // s, w
-        [r.geometry.viewport.northeast.lat, r.geometry.viewport.northeast.lng], // n, e
-      ],
-      raw: r,
-    }));
+  static async loadGeocoder(
+    options: ProviderOptions = {},
+  ): Promise<google.maps.Geocoder> {
+    return new Loader({
+      // The toString() is necessary to address a type mismatch -
+      // key can be string | number | boolean, but Google Maps
+      // expects it to be a string.
+      apiKey: options.params?.key?.toString() || '',
+      version: 'weekly',
+      // If a region is not specified, geocoding requests will bias toward the United States.
+      region: options.params?.region?.toString(),
+    })
+      .load()
+      .then((google) => {
+        return new google.maps.Geocoder();
+      });
+  }
+
+  endpoint({ query }: EndpointArgument): never {
+    throw new Error('Method not implemented.');
```

@darrenklein You were on the right track here. If a method is declared as abstract in the parent, there's no way around it. I just think it's best practice to `throw` in these cases instead of simply returning a string in order to prevent any future errors from being "swallowed".

---

Nope, this sounds good to me! I didn't know what the standard practice would be in this sort of situation, but that makes sense.